# Programmation fonctionnelle avancée (Scala, Apache Spark)

**Auteur** : Djibril MBOUP   
**Role** : Data Scientist à Atos Senegal  
**E-mail** : mboupdjibril.m2itic@gmail.com - mboup.djibril@ugb.edu.sn 
 

## Objectifs
Dans ce cours nous allons :

* Voir les bases de la programmation fonctionnelle en Scala,
* Manipuler les concepts avancées de collections, classes, higher-order functions, pattern matching, mapReduce etc.
* Developper des programmes paralleles avec Apache Spark pour le pré-traitement des données dans un cluster Hadoop
* Chargement, Transformation des données avec Spark Dataframe, SQL et Datasets.

## Le language Scala

Scala combine la programmation orientée objet et fonctionnelle dans un langage concis et de haut niveau. Les types statiques de Scala aident à éviter les bogues dans les applications complexes, et ses runtimes JVM et JavaScript vous permettent de créer des systèmes hautes performances avec un accès facile à d'énormes écosystèmes de bibliothèques.


## Scala REPL (Read-Eval-Print-Loop)

Scala peut fonctionner comme une calculatrice en permettant de :

* Lire une expression saisie dans le prompt, 
* Evaluer avec le compilateur Scala,
* Afficher dans le prompt.
* Attendre (en boucle) d'une nouvelle expression  

Si Scala est installé dans votre système, vous pouvez demarrer scala dans votre terminal.



En tapant 1 + 2, vous invoquez en fait une méthode nommée + définie dans la classe **Int**, puisque Scala en un lagange pure OO

In [1]:
1 + 2

res0: Int = 3

In [2]:
res0 * 3

res1: Int = 9

In [3]:
(res1 - res0)

res2: Int = 6

In [4]:
println("Hello, world!")

Hello, world!


La sortie println passe la chaine de caractère à la sortie standard, similaire à System.out.println de Java.

### Les variables  

Scala a deux type de variables, **val** et **var**

* **val** est similaire à une variable final dans Java. Une fois initialisée, la variable ne peut plus etre modifié.   
* Quant à **var** , il permet de dèclarer une variable non-final variable comme dans Java. Une variable dèclarée avec **var** peut change de valeur durant sa durée de vie.


In [5]:
val msg = "Hello, world!"

msg: String = "Hello, world!"

In [6]:
val msg2: java.lang.String = "Hello again, world!"

msg2: String = "Hello again, world!"

In [ ]:
val msg3: String = "Hello yet again, world!"

In [6]:
msg = "Goodbye cruel world!"

cmd6.sc:1: reassignment to val
val res6 = msg = "Goodbye cruel world!"
               ^Compilation Failed

: 

In [6]:
msg4 = "Variable failed due to the missing of val or var"

cmd6.sc:1: not found: value msg4
val res6 = msg4 = "Variable failed due to the missing of val or var"
           ^Compilation Failed

: 

In [7]:
var greeting = "Hello, world!"

greeting: String = "Leave me alone, world!"

In [8]:
greeting = "Leave me alone, world!"

In [9]:
println(greeting)

Leave me alone, world!


### Blocks

In [10]:
println({
  val x = 1 + 2
  x + 3
})

6


In [11]:
val bloc = {
  val x = 1 + 2
  x + 3
}
bloc

bloc: Int = 6
res10_1: Int = 6

### Controle Structures

### If expressions

In [12]:
val note = 10

if (note >= 10) 
    print("Pass")
else 
    print("Ajournée")


Pass

note: Int = 10

In [13]:
 val note = 10

if (note >= 10) {
    println("Pass")
    print("Moyenne = " +note)
}
else 
    print("Ajournée")

(console):7:1 expected end-of-input
else 
^

: 

#### Case match

In [14]:
import scala.util.Random

val x: Int = Random.nextInt(10)

x match {
  case 0 => "zero"
  case 1 => "one"
  case 2 => "two"
  case _ => "other"
}

import scala.util.Random


x: Int = 7
res13_2: String = "other"

In [15]:
val cmd = "stop"
cmd match {
    case "start" | "go" => println("starting")
    case "stop" | "quit" | "exit" => println("stopping")
    case _ => println("doing nothing")
}

stopping


cmd: String = "stop"

#### Boucle while / do while

In [16]:
val text = "Bienvenue dans le monde de Scala ! "
val tabString = text.split(" ")
tabString

text: String = "Bienvenue dans le monde de Scala ! "
tabString: Array[String] = Array(
  "Bienvenue",
  "dans",
  "le",
  "monde",
  "de",
  "Scala",
  "!"
)
res15_2: Array[String] = Array(
  "Bienvenue",
  "dans",
  "le",
  "monde",
  "de",
  "Scala",
  "!"
)

In [17]:
var i = 0
while (i < tabString.length) {
    println(tabString(i))
    i += 1
}

Bienvenue
dans
le
monde
de
Scala
!


i: Int = 7

In [18]:
var i = 0
do {
     println(tabString(i))
     i+=1
} while ( i < tabString.length)

Bienvenue
dans
le
monde
de
Scala
!


i: Int = 7

### For et Foreach

In [19]:
for (w <- tabString)
    println(w)

Bienvenue
dans
le
monde
de
Scala
!


In [20]:
for (i <- 0 to tabString.length - 1)
    println(tabString(i))

Bienvenue
dans
le
monde
de
Scala
!


Afficher les nombres impaires et 1 à 100

In [21]:
val n = 100
println("Liste des nombres impaires")
for (i <- 1 to n)
    if (i % 2 != 0 )
        print(i + " ")

Liste des nombres impaires
1 3 5 7 9 11 13 15 17 19 21 23 25 27 29 31 33 35 37 39 41 43 45 47 49 51 53 55 57 59 61 63 65 67 69 71 73 75 77 79 81 83 85 87 89 91 93 95 97 99 

n: Int = 100

Boucle for sur un map

In [22]:
val names = Map("Prenom" -> "Macky", "Nom" -> "Sall")

for ((k,v) <- names) 
    //println(s"key: $k, value: $v")
    println("key: "+k+", value: "+v)

key: Prenom, value: Macky
key: Nom, value: Sall


names: Map[String, String] = Map("Prenom" -> "Macky", "Nom" -> "Sall")

#### For comprehension (for/yield)

In [23]:
val fruits = Array("banana", "pomme", "orange")
val upper = for (e <- fruits) yield e.toUpperCase

fruits: Array[String] = Array("banana", "pomme", "orange")
upper: Array[String] = Array("BANANA", "POMME", "ORANGE")

In [24]:
val lengths = for (e <- fruits) yield {
                        e.length
}

lengths: Array[Int] = Array(6, 5, 6)

In [25]:
val newlist = for (e <- fruits ) yield {
        if(e.contains("a")) 
            e
}

newlist: Array[Any] = Array("banana", (), "orange")

In [26]:
val nombres = List(1, 2, 3, 5)
val carre = for (e <- nombres) yield e*e

nombres: List[Int] = List(1, 2, 3, 5)
carre: List[Int] = List(1, 4, 9, 25)

Foreach permet de lire de collections (List, set, array ...). Il prend souvent en argument une function

In [27]:
upper

res26: Array[String] = Array("BANANA", "POMME", "ORANGE")

In [28]:
upper.foreach(w => print(w + " "))

BANANA POMME ORANGE 

In [29]:
upper.foreach(println)

BANANA
POMME
ORANGE


In [ ]:
/*
sum = 0
for x <- tab
sum = sum + x
*/

In [30]:
val liste = List(1, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20)
var sum = 0
liste.foreach(sum += _) // x => sum+=x
println("Somme " + sum)

Somme 111


liste: List[Int] = List(1, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20)
sum: Int = 111

### Functions 


#### Fonction anonyme

In [31]:
(x: Int) => x + 1  //

res30: Int => Int = ammonite.$sess.cmd30$Helper$$Lambda$2614/0x0000000800c44840@172a06c3

In [32]:
res30(12)

res31: Int = 13

In [33]:
val incr = (x: Int) =>  x + 1 

incr: Int => Int = ammonite.$sess.cmd32$Helper$$Lambda$2621/0x0000000800c49040@2dc536b

In [34]:
incr(2)

res33: Int = 3

In [35]:
val add = (x: Int, y: Int) => x + y
println(add(1, 2))

3


add: (Int, Int) => Int = ammonite.$sess.cmd34$Helper$$Lambda$2627/0x0000000800c4c840@4bb10bbf

In [36]:
val getValue = () => 42

getValue: () => Int = ammonite.$sess.cmd35$Helper$$Lambda$2635/0x0000000800c51040@10e2cb3a

In [37]:
println(getValue()) 

42


### Keyword : def 

In [38]:
def max1(x: Int, y: Int): Int = {
        if (x > y) x
        else y
    }

defined function max1

In [39]:
max1(21, 18)

res38: Int = 21

In [40]:
def max2(x: Int, y: Int): Int = 
        if (x > y) x
        else y
    

defined function max2

In [41]:
max2(21, 18)

res40: Int = 21

In [42]:
def max3(x: Int, y: Int) = 
        if (x > y) x
        else y

defined function max3

In [ ]:
max3(21, 18)

In [ ]:
val maximum = (x: Int, y: Int) =>  
                if (x > y) x 
                else y

In [ ]:
maximum(21, 18)

In [43]:
def getSquareString(x: Double): String = {
  val square = x * x
  square.toString
}

defined function getSquareString

In [44]:
println(getSquareString(12))

144.0


In [45]:
def greet() = println("Hello, world!")

defined function greet

In [ ]:
greet()

In [ ]:
def bye() = println("Bye bye, Tomorrow!")
bye()

### Exception try/catch/finally
Similaire à Java


In [46]:
val throwsException = throw new RuntimeException  // Nothing

: 

In [47]:
val aPotentialFailure = try {
    throw new RuntimeException
  }catch {
    case e: Exception => "I caught an exception"
  }finally {
    println("some logs")
  }

some logs


aPotentialFailure: String = "I caught an exception"

## Scala Script

**Demo** sur IntelliJ ou VSCODE

Copier le code ci-dessous sur votre éditeur préféré :

```object HelloWorld {
        def main(args: Array[String]){
        println("Hello, world!")
        }
    }
```


## Exercice à faire 15 mn

**Exercice 1 :**  
Ecrire un programme scala qui permet de calculer le pgcd de deux nombre donnés en argument de ligne de commande.

**Exercice 2 :**  
Suite de Fibonnacci,  

F(0) = 1  
F(1) = 1  
F(n) = F(n-1) + F(n-2)

Ecrire un programme recusrsive en Scala qui permet de calculer la suite de Fibonnacci

## Exercice à Rendre

**Exercice 3 :**
Ecrire en Scala un programme qui prend en entrée un entier `n` et retourne son écriture en base 2.
 
 Exemple :  
 toBase2(2) -> 10  
 toBase2(7) -> 111
 
 **Exercice 4**:  
 Ecrire un programme en Scala qui permet de faire la recherche binaire dans une liste de nombres d'entier en utilisant la méthode divisé pour régner.
 

### Rappel

Les classes et les objets sont les composants fondamentaux de la programation orientée objet.

Le concept d’utilisation de **classes** et d’**objets** consiste à encapsuler l’état et le comportement dans une seule unité de programmation. Les objets sont similaires aux objets du monde réel. Par exemple, nous pouvons créer un objet voiture, qui aura des propriétés telles que la vitesse et la couleur; et un comportement comme: accélérer et freiner.  

![](images/oo.png)

In [48]:
class User

defined class User

Instanciation d'un objet de la classe User avec **new**

In [49]:
val user1 = new User

user1: User = ammonite.$sess.cmd47$Helper$User@6f6f2267

In [50]:
user1

res49: User = ammonite.$sess.cmd47$Helper$User@6f6f2267

#### classe Point

Création de la classe Point avec un constructeur qui a deux argument, les données membres sont accessibles et modifialbes grace au mot clé **var**. Le mot-clé **override** nous permet de surcharger la methode **toString** de la classe mère **Object**.

In [51]:
class Point(var x: Int, var y: Int) {

  def move(dx: Int, dy: Int): Unit = {
    x = x + dx
    y = y + dy
  }

  override def toString: String =
    s"($x, $y)"
}

defined class Point

In [52]:
val p1 = new Point(2, 3)
val p2 = new Point(5, 9)
println(p1.x)
println(p1.y)
println(p2)

2
3
(5, 9)


p1: Point = (2, 3)
p2: Point = (5, 9)

**Erreur de compilation de compilation**: Pas de constructeur avec un seul argument

In [52]:
val p3 = new Point(2)

cmd52.sc:1: not enough arguments for constructor Point: (x: Int, y: Int): cmd52.this.cmd50.Point.
Unspecified value parameter y.
val p3 = new Point(2)
         ^Compilation Failed

: 

### Constructeurs

Création de la classe Point avec un constructeur ayant des valeurs par défaut.

In [53]:
class Point(var x: Int = 0, var y: Int = 0)

val p0 = new Point  
val p1 = new Point(1) // par 1 sera par defaut affecté à x
println("p1.x = " + p1.x) 

p1.x = 1


defined class Point
p0: Point = ammonite.$sess.cmd52$Helper$Point@52e649e6
p1: Point = ammonite.$sess.cmd52$Helper$Point@23824bd5

In [54]:
val p2 = new Point(y = 2)
println(p2.y)  

2


p2: Point = ammonite.$sess.cmd52$Helper$Point@61e11da

In [55]:
class Point(var x: Int = 0, var y: Int = 0) {

  def move(dx: Int, dy: Int): Unit = {
    x = x + dx
    y = y + dy
  }

  override def toString: String =
    s"($x, $y)"
}

defined class Point

In [ ]:
val p0 = new Point
val p1 = new Point(1)
val p2 = new Point(2,6)

In [ ]:
p2.move(1,0)

In [ ]:
println(p2)

In [ ]:
var p3 = new Point(2,1)
p3 = p2

### Membres privates et Getter/Setter

Les données membres _x et _y ne sont accessibles qu'appartir du getteur et du setteur

In [56]:
class Point {
  private var _x = 0
  private var _y = 0
  private val bound = 100
  
    def this(x : Int, y: Int) = {
      this(); // appel du constructeur par defaut
      _x = x
      _y = y
    }

  // Definition d'un getteur 
  def x = _x
  
  // Definition d'un setteur
  def x_= (newValue: Int): Unit = {
    if (newValue < bound) _x = newValue else printWarning
  }
    
    

  def y = _y
  def y_= (newValue: Int): Unit = {
    if (newValue < bound) _y = newValue else printWarning
  }

  private def printWarning = println("WARNING: Out of bounds")

  override def toString: String =
    s"($x, $y)"
}

defined class Point

Dans cette classe, il n y a pas de constructeur explicite, c'est le constructeur par defaut qui est appelé

In [57]:
val p0 = new Point

p0: Point = (0, 0)

In [58]:
// Le setteur est appliqué
p0.x = 1
p0.y = 5

In [59]:
// Le getteur est appliqué
p0.x

res58: Int = 1

In [60]:
println(p0)

(1, 5)


In [61]:
val p1 = new Point(1,1)

p1: Point = (1, 1)

In [62]:
val p2 = new Point
p2.x = 54
p2.y = 101 

p2: Point = (54, 0)

### Accés Publiques/privates 
Les parametres du constructeur sont public avec **var** ou **val**. Cependant, **val** ne permet pas de modifier la variable (immutable).

In [ ]:
class Point(val x: Int, var y: Int)

In [ ]:
val p = new Point(1, 2)

In [ ]:
p.y = 5  // modifiable à cause de mot-clé var

In [ ]:
p.x = 3  // non modifiable puisqu'il est défini avec val

In [ ]:
p.x

**Attention** Si le constructeur est déclaré sans **val** ou **var**, les paramètres son privates.

In [ ]:
class Point(x: Int, y: Int)
val point = new Point(1, 2)

Nous ne pouvons pas accéder à cette variable parcequ'il est private

In [ ]:
point.x

### Classe rationnelle

In [ ]:
class Rational(x: Int, y : Int){
     
    require(y!=0, "Denominator must be nonzero")
    // Definition d'un constructur à partir du premier constructeur
    
    def this(x: Int) = this(x,1)
    
    // cacul du pgcd de deux nombre
    private  def gcd(a:Int, b: Int): Int = if (b==0) a else gcd(b, a%b)
    
    val g =gcd(x,y)
  
    def numer = x / g
    
    def denom = y / g

    def + (that : Rational) = new Rational(numer * that.denom + denom * that.numer,
          denom * that.denom)
  
    def unary_- : Rational = new Rational(-numer, denom)
    
    def - (that: Rational) = this + -that
    
    def < (that: Rational) = this.numer * that.denom < that.numer * this.denom
    
    def > (that: Rational) = that < this
    
    def <= (that: Rational) = (this < that) || (this == that)
    
    def >= (that: Rational) = (this > that) || (this == that)
  
    def max(that : Rational) = if(this < that) that else this
    
    
    override def toString: String = x + "/" + y
}

In [ ]:
val a = new Rational(2)

In [ ]:
val r = new Rational(1, 2)
val q = new Rational (2,3)
val z = new Rational(7, 8)

In [ ]:
q + r

In [ ]:
z - r - q // z.-(r.-(q))

In [ ]:
val b = z - q + r

In [ ]:
r.max(q)

## Exercice 1 (A Rendre)

S'inspirer vous de la classe Rationnel pour définir une classe Vecteur2D en sursargeant les operateurs :  
 * \+ : pour faire la somme de deux vecteurs 
 * \- : pour faire la soustraction de deux vecteurs
 * \* : le produit scalaire de deux vecteurs 
 * == pour verifier l'ègalité de deux vecteurs  
 
En definissant les methodes :  
 * norm : pour calculer la norme d'un vecteur
 * dist : pour calculer la distance entre deux vecteur
 

### case class

**case class** est utilisé pour representé souvent des structures de données complexes.

In [ ]:
case class Book(title : String, authors :String, isbn: String)
val livre = Book("Une Si Longue Lettre","Mariama Ba", "2842612892")

In [ ]:
livre.title

In [ ]:
case class Message(sender: String, recipient: String, body: String)
val message1 = Message("guillaume@quebec.ca", "jorge@catalonia.es", "Ca va ?")

println(message1.sender)  

In [ ]:
message1.sender = "travis@washington.us" //unmutable

In [ ]:
val message2 = Message("jorge@catalonia.es", "guillaume@quebec.ca", "Com va?")
val message3 = Message("jorge@catalonia.es", "guillaume@quebec.ca", "Com va?")
message2 == message3

### Object singleton instance

Object est une classe qui a exactement une instance. Il est créé           paresseusement lorsqu'il est référencé, comme un lazy val.

En tant que valeur de niveau supérieur, un objet est un singleton.

En tant que membre d'une classe englobante ou en tant que valeur locale, il se comporte exactement comme un lazy val.

In [63]:
object Box

defined object Box

Definition de la classe Logger dans le package logging

La méthode `info` peut être importée dans n'importe quel programme. La création de méthodes utilitaires comme celle-ci est un cas d'utilisation courant pour les objets singleton.

Le singleton object est utilisé comme d'entrée d'un programme scala

In [ ]:
object Hello {
    def main(args: Array[String]) = {
        println("Hello, world")
    }
}

### Heritage

L’héritage est un concept fondamental des languages Orientées Objets qui permet de réutiliser du code de la classe mère.

Le polymorphisme permet de "surcharger" les méthodes de la classe mère pour redéfinir leurs comportements sans changer leur signature.

In [ ]:
class Animal (var name: String, var age: Int) {
    // (2) auxiliary constructor
    def this (name: String) {
        this(name, 0) //appel du constructeur par defaut
    }
    
    override def toString = s"$name is $age years old"
}

In [ ]:
class Dog (name: String) extends Animal (name) {
    println("Dog constructor called")
}

In [ ]:
class Cat (name: String) extends Animal (name) {
    println("Cat constructor called")
}

In [ ]:
val an = new Animal("Bobby", 2) 

In [ ]:
val dog = new Dog("Bobby")

In [ ]:
val cat = new Cat("Chat")

### Traits

Trait est très similaire à la notion d'interface dans Java, à la seul difference qu'il peut contenir des methods concrètes.

In [ ]:
trait Figure{
    //def dessiner()
    def calculerPerimetre() : Double
    def CalculerSurface() : Double
}

In [ ]:
class Rectangle (var longueur : Double, var largeur : Double) extends Figure {
    
    def calculerPerimetre() = (longueur + largeur)*2
    def CalculerSurface() = longueur*largeur
}

In [ ]:
val r = new Rectangle(2.0, 5.0)

In [ ]:
r.calculerPerimetre()

Definissons un iterateur à partir de Trait générique

In [ ]:
trait Iterator[A] {
  def hasNext: Boolean
  def next(): A
}

Un iterateur d'entier à partir du Trait Iterator

In [ ]:
class IntIterator(to: Int) extends Iterator[Int] {
  private var current = 0
  override def hasNext: Boolean = current < to
  override def next(): Int = {
    if (hasNext) {
      val t = current
      current += 1
      t
    } else 0
  }
}

In [ ]:
val iterator = new IntIterator(10)
iterator.next()

In [ ]:
iterator.next()  

In [ ]:
iterator.next()  

In [ ]:
iterator.hasNext

Une autre definition d'une Figure rectangulaire en considèrant les points d'intersection des ses cotés

In [ ]:
class Point(val x: Int, val y: Int)

trait Rectangular {
    def topLeft: Point
    def bottomRight: Point
    def left = topLeft.x
    def right = bottomRight.x
    def width = right - left
    // autres methodes ...
}

In [ ]:
class Rectangle(val topLeft: Point, val bottomRight: Point) extends Rectangular {
    // autres methods
}

In [ ]:
val rect = new Rectangle(new Point(1, 1), new Point(10, 10))

In [ ]:
rect.left

In [ ]:
rect.right

In [ ]:
rect.width

### Abstract class

Scala a également un concept de **classe abstraite** similaire à la classe abstraite de Java. Mais comme les traits sont si puissants, vous avez rarement besoin d'utiliser une classe abstraite. En fait, vous n'avez besoin d'utiliser une classe abstraite que lorsque:

* Vous souhaitez créer une classe de base qui nécessite des arguments de constructeur
* Votre code Scala sera appelé à partir du code Java

In [ ]:
abstract class Pet (name: String) {
    val greeting: String
    
    var age: Int
    
    def sayHello { 
        println(greeting) 
    }
    override def toString = s"I say $greeting, and I'm $age"
}

In [ ]:
class Dog (name: String) extends Pet (name) {
    val greeting = "Woof"
    var age = 2
}
class Cat (name: String) extends Pet (name) {
    val greeting = "Meow"
    var age = 5
}

In [ ]:
val dog = new Dog("Fido")

val cat = new Cat("Morris")

dog.sayHello

cat.sayHello
    

In [ ]:
println(dog)

println(cat)

cat.age = 10

println(cat)

### Exemple

Definition d'un ensemble d'entier à partir de ces propriétés

In [ ]:
abstract class IntSet {
    
  def contains(x : Int) : Boolean
    
  def incl(x: Int) : IntSet
    
  def union(other : IntSet) : IntSet
}

In [ ]:
class NonEmpty (elem: Int, left: IntSet,
                right: IntSet ) extends IntSet {
    
  def contains(x: Int): Boolean =
        if (x < elem) left contains x
        else if (x > elem) right contains x
        else true

  def incl(x: Int): IntSet =
        if(x < elem) new NonEmpty(elem, left incl x,  right)
        else if (x > elem) new NonEmpty(elem, left, right incl x )
        else this

  override def union(other: IntSet): IntSet =
    ((left union right) union other) incl elem
  
  override def toString: String =
    "{" + left + elem + right + "}"
}

In [ ]:
class Empty extends IntSet  {
    
  def contains(x :Int) : Boolean= false
    
  def incl(x: Int) : IntSet = new NonEmpty (x, new Empty, new Empty)
    
  def union(other : IntSet) = other
    
  override def toString: String = "."
    
}

In [ ]:
val t1  = new NonEmpty(1, new Empty, new Empty)

In [ ]:
val t2 = t1 incl 3 // t1.incl(3)

In [ ]:
 t2 contains 2 // t2.contains(2)

In [ ]:
t2.contains(2)

In [ ]:
t2 contains 3

In [ ]:
val t3 = ((t2 incl 9) incl 4) incl 15

In [ ]:
val t4 = (t1 incl 2) incl 7

In [ ]:
val t5 = t3 union t4

## Polymorphysme

In [ ]:
class Animal

In [ ]:
class Dog extends Animal

In [ ]:
val aDog: Animal = new Dog // subtyping polymorphism

In [ ]:
trait Carnivore {
    def eat(a: Animal): Unit
}

In [ ]:
class Crocodile extends Animal with Carnivore {
    override def eat(a: Animal): Unit = println("crunch!")
  }

In [ ]:
val aCroc = new Crocodile

In [ ]:
aCroc.eat(aDog)

In [ ]:
aCroc eat aDog // natural language

### Classe générique

In [ ]:
abstract class Set[A] {
    
  def incl(a: A): Set[A]
  def contains(a: A): Boolean
    
  def union(other : A) : A
  def intersect(other : A) : A
  def diff(other : A) : A
    
}

class Empty[A] extends Set[A] {
  //…
}

class NonEmpty[A](elem: A, left: Set[A], right: Set[A]) extends Set[A] {
  //…
}

### Exercice 2 (A Rendre)

Reprendre les methodes de la classe Intset dans la classe générique.

### Package 

**Importer les classes d'un package Scala**  
L'importation des classes en Scala est très similaire à celle de Java.

Supposons qu'on a un package qui contient les classes Animal, Personne, Voiture.

**Importation des classes Java**

Aliasing pour éviter le conflit des noms

In [ ]:
import java.util.{Date => utilDate}
import java.sql.{Date => sqlDate}

In [ ]:
def run(): Unit = {
              val dt: utilDate = new utilDate()
              val dtSql: sqlDate = new sqlDate(System.currentTimeMillis())
              println(s"Today is $dt !")
              println(s"Today is $dtSql !")
}

In [ ]:
run()

#### Creation de Package  package 

In [ ]:
// Demo IntelliJ

### Interoperabilité avec Java

In [ ]:
//... Voir Documentation

### Travail à faire
Faire les exercices 1 et 2 de ce notebook en sous la forme chap2_execice1.scala et chap2_execice1.scala.
  
N'oublier pas de zipper en mettant votre_nom_master.zip sur le mail ci-dessous :


mboupdjibril.m2itic@gmail.com

# Scala APIs: String, Collections, Lists, Arrays, Map, Set ...

Scala possede un ensemble de structures et d'API permettant de manipuler et de traiter les données. 

### Types 

![](images/types.svg)

**Any** est le super-type. Il defini par défaut un ensemble de methodes : equals, hashCode, and toString.
Il a deux principales sous-classes: AnyVal et AnyRef.
* AnyVal :
* AnyRef :

### Casting

![](images/casting.svg)

### La classe String

In [ ]:
//Obtenir le nom de la classe
"Hello, world".getClass.getName

In [ ]:
"hello".foreach(println)

In [ ]:
// concatenation
val s = "Hello" + " world" 

In [ ]:
//longueur d'une chaine de caractère
s.length

Filtrer un caractère 

In [ ]:
val result = s.filter(_ != 'l')
// est equivalent à s.filter(x  => x != 'l') 

In [ ]:
result

In [ ]:
"bonjour".capitalize

In [ ]:
"bonjour".take(2).capitalize

In [ ]:
"bonjour".drop(3).take(2).capitalize

In [ ]:
"bonjour" drop(3) take(2) 

In [ ]:
val s = "salut"
s.toUpperCase

In [ ]:
val s1 = "hello"
val s2 = "Hello"

In [ ]:
s1 == s2

In [ ]:
s1.toUpperCase == s2.toUpperCase

In [ ]:
s1.equalsIgnoreCase(s2)

In [ ]:
val foo = """Ceci est
une chaine 
multiline"""

In [ ]:
val speech = """Four score and
|seven years ago""".stripMargin

In [ ]:
val speech = """Four score and
#seven years ago""".stripMargin('#')

In [ ]:
val s = """This is known as a
|"multiline" string
|or 'heredoc' syntax.""". stripMargin.replaceAll("\n", " ")

In [ ]:
"hello".toLowerCase

In [ ]:
"Ceci est une Phrase coupée moT par mot et convertie en minuscule".split(" ").map(_.toLowerCase)

In [ ]:
// split permet de découper une chaine selon le delimiter
val languages = " Java, Python, C++, Scala, Php, Fortran, Julia "
languages.split(",")

In [ ]:
// trim supprime les chaines vides en debut et fin de ligne
languages.split(",").map(_.trim)

In [ ]:
// Avec les expression régulières
languages.split("\\s+")

In [ ]:
//Filtrer la lettre et convertir en majuscule 
val upper = "hello world".filter(_ != 'l').map(_.toUpper)

### Trouver un pattern dans la chaine de caractère

In [ ]:
// Definir un pattern pour extraire les nombres
val numPattern = "[0-9]+".r 

In [ ]:
val address = "123 Main Street Suite 101"

On veut extraire le numero de rue de la ville dans une adresse

In [ ]:
val match1 = numPattern.findFirstIn(address)

In [ ]:
val matches = numPattern.findAllIn(address)
matches.foreach(println)

In [ ]:
matches

In [ ]:
//ou bien de le convertir en Array
val matches = numPattern.findAllIn(address).toArray

In [ ]:
val result = numPattern.findFirstIn(address).getOrElse("no match")

In [ ]:
val address2 = "Rue Carnot x Alpha"

In [ ]:
numPattern.findFirstIn(address2).getOrElse("no match")

In [ ]:
val address = "123 Main Street".replaceAll("[0-9]", "x")

In [ ]:
"123".replaceFirst("[0-9]", "x")

In [ ]:
val regex = "H".r
val result = regex.replaceFirstIn("Hello world", "P")

In [ ]:
"120".toInt

### Collections

Les collections dans Scala sont de deux types : 
* collections mutables : Elles peuvent être mise à jour ou étendue sur place. Ce qui veut dire que vous pouvez modifier, ajouter ou supprimer des éléments d'une collection en tant qu'effet secondaire.  Dans scala, nous les avons dans le package ``scala.collection.mutable``
* collections immuables:  Elles, en revanche, ne changent jamais. Vous avez toujours des opérations qui simulent des ajouts, des suppressions ou des mises à jour, mais ces opérations renverront dans chaque cas une nouvelle collection et laisseront l'ancienne collection inchangée. Ils se trouvent dans le package ``scala.collection.immutable``

 package ``scala.collection``

![](images/collections-diagram.svg)

package ``scala.collection.immutable``

![](images/collections-immutable-diagram.svg)

package ``scala.collection.mutable``

![](images/collections-mutable-diagram.svg)

### List

Une liste est une collection immuable, elle permet d'ordoner ses éléments et garder les duplicats. Les listes en Scala sont differents de celles conçues en Java (``ArrayList``). Selon le paradigme fonctionnelle, une liste est implémenté sous la forme d'une liste chainée avec les principales methodes suivantes :
* **head**
* **tail**
* **isEmpty**

#### creation d'une Liste en scala

In [ ]:
val list = 1 :: 2 :: 3 :: Nil

In [ ]:
val list = List(1, 2, 3)

In [ ]:
val x = List(1, 2.0, 33D, 4000L)

In [ ]:
val x = List[Number](1, 2.0, 33D, 4000L)

In [ ]:
val x = List.range(1, 10)

In [ ]:
val x = (1 to 10).toList

In [ ]:
(1 to 10)

In [ ]:
val x = List.range(1, 10)

In [ ]:
val x = List.range(0, 10, 2)

In [ ]:
val x = List.fill(3)("foo")

In [ ]:
val x = List.tabulate(5)(n => n * n)

In [ ]:
"hello".toList

#### Ajouter et supprimer des elements dans une liste

In [ ]:
// ajout debut
val x = List(1,2)
val y = 0::x

In [ ]:
y

In [ ]:
var x = List(1, 2)
x = 0::x

In [ ]:
// ajout equivalent à append
val x = List(1,2)
val y = 0+:x

In [ ]:
val z = y:+3

Une liste est immuable, vous ne pouvez donc pas en supprimer des éléments, mais vous pouvez filtrer.

In [ ]:
val l = (1 to 10).toList 

In [ ]:
val newList = l.filter( _ > 4) // x => x > 4

In [ ]:
val newList = l.filter( _ % 2 == 0)

#### Mutable List 

In [ ]:
import scala.collection.mutable.ListBuffer
var fruits = new ListBuffer[String]()
// ajout
fruits += "Apple"
fruits += "Apple"
fruits += "Orange"

In [ ]:
// ajouter plusieurs elements
fruits += ("Strawberry", "Kiwi", "Pineapple")

#### Affichage

In [ ]:
fruits.foreach(println)

In [ ]:
fruits.remove(2)

In [ ]:
// supprimer un element
fruits -= "Apple"

In [ ]:
// supprimer plusieurs elements
fruits -= ("Banana", "Orange")

In [ ]:
val fruitsList = fruits.toList

#### Concatenation

In [ ]:
val a = List(1,2,3)
val b = List(4,5,6)

In [ ]:
val c = a ++ b

In [ ]:
val c = a ::: b

#### Afficher les element d'une list avec take, takewhile

In [ ]:
val nums = (1 to 10).toList

In [ ]:
nums.take(1)

In [ ]:
nums.take(3)

In [ ]:
nums.takeWhile(_ < 5)

In [ ]:
val names = List("aly","pape", "modou","fatou")

In [ ]:
names.takeWhile(_.length <= 3)

In [ ]:
names.takeWhile(_.length < 5)

In [ ]:
names.takeWhile(_.length < 6)

#### Map, flatMap

In [ ]:
// carrée des nombre
val maplist = List(2, 4, 5, -6) map ( x=> x * x)
// maplist.map(x=> x * x)

In [ ]:
val flatmap = List(List(2,4), List(6,8)) flatMap(x => x.map(x => x * x))

#### Map Reduce

In [ ]:
val nombres = List(2, 8, 5, 6, 4, 7, 3)
val res = nombres.reduce((x, y) => x max y)

In [ ]:
// map reduce pour calculer la moyenne
 val pairs = nombres.map(x => (x,1))

In [ ]:
val tuple= (1, 2)

In [ ]:
val res = pairs.reduce((a,b) => (a._1 + b._1, 
                                            a._2 + b._2 ) ) 

In [ ]:
println("Moyenne = "+ res._1/res._2.toFloat) 

In [ ]:
// Ecriture condensée
val res = nombres.map(x => (x,1)).reduce((a,b) => ( a._1 + b._1, a._2 + b._2 ))

In [ ]:
println("Moyenne = "+ res._1/res._2.toFloat) 

### Array

Un Array en scala est un type mutable; l'odre et la duplication des élements est gardé. 

In [ ]:
val numbers: Array[Int] = Array[Int](1, 2, 3, 4, 5, 1, 2, 3, 3, 4, 5)

#### Modifier un élément dans un Array

In [ ]:
numbers(5)

In [ ]:
numbers(5) = 6

In [ ]:
numbers(7) = 7

In [ ]:
numbers(7)

#### Afficher les elements d'un Array

In [ ]:
println("The full array is: ")
for (elem <- numbers) {
    print(" " + elem)
}

In [ ]:
numbers.foreach(println)

#### Somme des élements d'un tableau

In [ ]:
var total = 0;
for (i <- 0 to (numbers.length - 1)) {
    total = total + numbers(i)
}
println("Sum: = " + total)

In [ ]:
print("Sum := " +numbers.sum)

#### Minimum d'un Array

In [ ]:
var min = numbers(0)
for (i <- 1 to (numbers.length - 1)) {
    if (numbers(i) < min) min = numbers(i)
}
println("Min is: " + min)

In [ ]:
print("Min := " +numbers.min)

In [ ]:
print("Max := "+numbers.max )

#### ordonner les élements d'un tableau

In [ ]:
import scala.util.Sorting._

In [ ]:
quickSort(numbers)

In [ ]:
numbers

#### Tableau Multi-dimensionel

In [ ]:
import Array._

In [ ]:

var myMatrix = ofDim[Int](4, 4)
// build a matrix
for (i <- 0 to 3) {
    for (j <- 0 to 3) {
    myMatrix(i)(j) = j*i
    }
}

println();

// Affichage
for (i <- 0 to 3) {
    for (j <- 0 to 3) {
    print(" " + myMatrix(i)(j))
    }
println();
}

### ArrayBuffer : un Array de taille modifiable

In [ ]:
import scala.collection.mutable.ArrayBuffer

In [ ]:
var prenoms = ArrayBuffer[String]("Abdou")

#### Ajouter d'élement dans ArrayBuffer

In [ ]:
prenoms += "Demba"
prenoms += "Albert"

In [ ]:
prenoms+= ("Pape", "Fall")

In [ ]:
prenoms++= Seq("Modou", "Ibrahim")

#### Suppression d'élément dans ArrayBuffer

In [ ]:
prenoms-="Albert"

In [ ]:
prenoms-=("Pape", "Fall")

In [ ]:
prenoms.remove(2)

### Map

Un Map est une structure clé/valeur qui correspond à dictionnaire

In [ ]:
val states = Map("SN" -> "Senegal", "US" -> "United-States", "FR" -> "France")

Par défaut, un Map immuable est créé

mutable Map

In [ ]:
//import collection.mutable.Map

In [ ]:
var states = collection.mutable.Map("SN" -> "Senegal", "US" -> "United-States", "FR" -> "France")

In [ ]:
//Ajout dans mutable Map
states += ("MO" -> "Maroc")

In [ ]:
states += ("ML" -> "Mali", "CI" -> "Cote d'Ivoire")

In [ ]:
for((k,v) <- states) println(s"$k, $v")

In [ ]:
states -= "FR"

In [ ]:
states -= ("ML", "US")

In [ ]:
states("SN") = "Pays de la Téranga"

In [ ]:
for((k,v) <- states) println(s"$k, $v")

In [ ]:
states.put("ML", "Mali")

In [ ]:
states.retain((k,v) => k == "SN")

In [ ]:
states.remove("SN")

In [ ]:
val s = states.getOrElse("SN", "No such state")

In [ ]:
import scala.collection.SortedMap

In [ ]:
val notes = SortedMap("Aly" -> 15, "Pape" -> 12, "Aicha" -> 17, "Fatou" -> 9, "Ibra" -> 7)

**LinkedHashMap** et **ListMap** pour garder l'ordre d'insertion des éléments

Lire la Documentation pour plus de détails 

### Set

Immutable Set

In [ ]:
val s1 = Set( 1, 2, 3, 4, 5, 1, 2, 3, 4, 5)

In [ ]:
val s2 = s1 + 5

In [ ]:
val s3 = s2 + (8, 7)

In [ ]:
val s4 = s3 ++ List(9, 4)

In [ ]:
var set = Set(1, 2, 3)

In [ ]:
set += 4

Mutable Set

In [ ]:
var set = scala.collection.mutable.Set[Int]()

Ajout

In [ ]:
set += 1

In [ ]:
set += (2, 3)

In [ ]:
set ++= Vector(4, 5)

In [ ]:
set.add(6)

In [ ]:
set.add(5)

In [ ]:
set.contains(5)

In [ ]:
set -= 1

In [ ]:
set -= (2, 3)

In [ ]:
set --= Array(4,5)

In [ ]:
var set = scala.collection.mutable.Set(1, 2, 3, 4, 5)

In [ ]:
set.retain(_ > 2)

SortedSet

In [ ]:
class Person (var name: String)

In [ ]:
import scala.collection.SortedSet

In [ ]:
val aliou = new Person("Aliou")
val cheikh = new Person("Cheikh")
val marc = new Person("Marc")
val you = new Person("You")

**Error de compilation :** la liste n'est pas ordonnée

In [ ]:
val s = SortedSet(aliou, marc, you, cheikh)

In [ ]:
class Person (var name: String) extends Ordered [Person]
{
    override def toString = name
    // return 0 if the same, negative if this < that, positive if this > that
    def compare (that: Person) = {
        if (this.name == that.name)
            0
        else if (this.name > that.name)
            1
        else
            -1
    }
}

In [ ]:
val aliou = new Person("Aliou")
val cheikh = new Person("Cheikh")
val marc = new Person("Marc")
val you = new Person("You")

In [ ]:
val ss = SortedSet(marc, you, cheikh, aliou)

**Queue**, permet dcreer une pile en scala  
Lire la Documentation pour plus de détails 

### Pattern matching

In [ ]:
import scala.io.StdIn.readLine

In [ ]:
print("Entrer un entier: ")
//val input = readLine() //A tester dans scala REPL

In [ ]:
val strx = "9"
val x = strx.toInt

In [ ]:

x match {
    case 1 => "un"
    case 2 => "deux"
    case _ => "superieur à deux"
}

In [ ]:
x match {
case 1 | 3 | 5 | 7 | 9 => println("Impaire")
case 2 | 4 | 6 | 8 | 10 => println("Paire")
}

In [ ]:
def generalSize(x: Any) = x match {
    case s: String => s.length
    case m: Map[_, _] => m.size
    case _ => -1
}

In [ ]:
generalSize("abc")

In [ ]:
generalSize(Map(1 -> 'a', 2 -> 'b'))

In [ ]:
trait Animal
    case class Chien(name: String) extends Animal
    case class Chat(name: String) extends Animal
    case object Perroquet extends Animal

In [ ]:
def determineType(x: Animal): String = x match {
    case Chien(moniker) => "Je suis un chien, je m'appelle " + moniker
    case _:Chat => "Je suis un chat"
    case Perroquet => "Je suis un perroquet"
    case _ => "Inconnu"
}

In [ ]:
println(determineType(new Chien("Rocky")))

In [ ]:
println(determineType(new Chat("Milou")))

In [ ]:
println(determineType(Perroquet))

In [ ]:
// somme des nombres paires d'une liste

def sommePaires(inputList: List[Int]): Int = inputList match {
   case Nil => 0 // liste vide
   case courant :: restant if courant % 2 == 0 => courant + sommePaires(restant)
   case _ :: restant => sommePaires(restant)
 }

In [ ]:
val liste = List(1, 2, 3, 4, 5, 6, 7, 8, 9)
sommePaires(liste)

Lire la Documentation pour plus de détails :  
https://docs.scala-lang.org/